File: frontend/lib/industry-market/market-series.ts
/**
 * Market-series growth and segmentation reconciliation (NOT persisted).
 */

import {
  abs,
  compare,
  difference,
  differsBeyond,
  div,
  isFilledDecimal,
  pct,
  round,
  subtract,
  sumDecimals,
  toFixedString,
  toNumber,
} from '@/lib/industry-market/decimal';
import type { MarketSegmentationRecord } from '@/lib/schemas/industry-market';

export type CagrReconciliation = {
  reported: string;
  calculated: string;
  difference: string;
  reconciles: boolean;
  message: string;
};

export type SegmentPercentageReconciliation = {
  parentMarketSeriesId: string;
  period: string;
  totalPercentage: string;
  reconciles: boolean;
  flags: string[];
};

/** Year-on-year growth: ((current - prior) / |prior|) × 100. */
export function calculateYoYGrowth(
  currentValue: string,
  priorValue: string,
  decimalPlaces = 2,
): string {
  if (!isFilledDecimal(currentValue) || !isFilledDecimal(priorValue)) return '';
  const priorCmp = compare(priorValue, '0');
  if (priorCmp === null || priorCmp === 0) return '';
  return pct(difference(currentValue, priorValue), abs(priorValue), decimalPlaces);
}

/**
 * CAGR over `periodCount` intervals: ((end / start) ^ (1 / periodCount) - 1) × 100.
 *
 * Fractional exponent uses `toNumber` for the root step; inputs and output remain decimal strings.
 */
export function calculateCagr(
  startValue: string,
  endValue: string,
  periodCount: number,
  decimalPlaces = 2,
): string {
  if (!isFilledDecimal(startValue) || !isFilledDecimal(endValue)) return '';
  if (periodCount <= 0) return '';
  const startCmp = compare(startValue, '0');
  if (startCmp === null || startCmp <= 0) return '';

  const ratio = div(endValue, startValue, 12);
  if (ratio === '') return '';

  const ratioNum = toNumber(ratio);
  if (ratioNum === null || ratioNum <= 0) return '';

  const cagr = (Math.pow(ratioNum, 1 / periodCount) - 1) * 100;
  if (!Number.isFinite(cagr)) return '';
  return toFixedString(String(cagr), decimalPlaces);
}

export function reconcileReportedVsCalculatedCagr(
  reported: string,
  calculated: string,
  tolerance = '0.5',
): CagrReconciliation {
  const hasReported = isFilledDecimal(reported);
  const hasCalculated = isFilledDecimal(calculated);

  if (!hasReported && !hasCalculated) {
    return {
      reported: '',
      calculated: '',
      difference: '',
      reconciles: false,
      message: 'Neither reported nor calculated CAGR is available.',
    };
  }

  if (!hasReported) {
    return {
      reported: '',
      calculated,
      difference: '',
      reconciles: false,
      message: 'Reported CAGR not entered — calculated value shown for reference only.',
    };
  }

  if (!hasCalculated) {
    return {
      reported,
      calculated: '',
      difference: '',
      reconciles: false,
      message: 'Calculated CAGR cannot be computed from available period values.',
    };
  }

  const delta = round(difference(reported, calculated), 2);
  const reconciles = !differsBeyond(reported, calculated, tolerance);

  return {
    reported,
    calculated,
    difference: delta,
    reconciles,
    message: reconciles
      ? 'Reported and calculated CAGR reconcile within tolerance.'
      : `Reported CAGR differs from calculated by ${delta} percentage points.`,
  };
}

/**
 * Sum sibling segment percentages for each parent market series and period.
 * Flags totals materially above 100% or with overlapping-definition concerns.
 */
export function reconcileSegmentPercentages(
  segments: MarketSegmentationRecord[],
  tolerance = '2',
): SegmentPercentageReconciliation[] {
  const groups = new Map<string, MarketSegmentationRecord[]>();

  for (const segment of segments) {
    if (!segment.parentMarketSeriesId.trim()) continue;
    const key = `${segment.parentMarketSeriesId}::${segment.period.trim()}`;
    const existing = groups.get(key) ?? [];
    existing.push(segment);
    groups.set(key, existing);
  }

  const results: SegmentPercentageReconciliation[] = [];

  for (const [key, group] of groups) {
    const [parentMarketSeriesId, period] = key.split('::');
    const percentages = group
      .map((segment) => segment.marketSharePercentage)
      .filter(isFilledDecimal);
    const totalPercentage =
      percentages.length > 0 ? sumDecimals(percentages) : '';

    const flags: string[] = [];
    if (group.some((segment) => !segment.parentMarketSeriesId.trim())) {
      flags.push('Segment missing parent market series.');
    }
    if (group.some((segment) => !isFilledDecimal(segment.marketSharePercentage))) {
      flags.push('One or more segments lack a market-share percentage.');
    }
    if (isFilledDecimal(totalPercentage)) {
      const totalVs100 = compare(totalPercentage, '100');
      if (totalVs100 !== null) {
        if (totalVs100 > 0 && !differsBeyond(totalPercentage, '100', tolerance)) {
          flags.push(`Segment percentages sum to ${totalPercentage}% — exceeds 100%.`);
        } else if (totalVs100 > 0) {
          flags.push(`Segment percentages sum to ${totalPercentage}% — materially above 100%.`);
        }
        const totalVsLowerBound = compare(totalPercentage, subtract('100', tolerance));
        if (totalVsLowerBound !== null && totalVsLowerBound < 0) {
          flags.push(`Segment percentages sum to ${totalPercentage}% — materially below 100%.`);
        }
      }
    }

    const reconciles =
      flags.length === 0 &&
      isFilledDecimal(totalPercentage) &&
      !differsBeyond(totalPercentage, '100', tolerance);

    results.push({
      parentMarketSeriesId,
      period,
      totalPercentage,
      reconciles,
      flags,
    });
  }

  return results;
}

File: backend/app/modules/industry_market/market_series.py
"""Market-series growth and segmentation reconciliation (NOT persisted)."""

from __future__ import annotations

import math
from typing import Any

from app.modules.industry_market import decimal_utils as dm


def calculate_yoy_growth(
    current_value: str,
    prior_value: str,
    decimal_places: int = 2,
) -> str:
    if not dm.is_filled(current_value) or not dm.is_filled(prior_value):
        return ""
    prior_cmp = dm.compare(prior_value, "0")
    if prior_cmp is None or prior_cmp == 0:
        return ""
    return dm.pct(dm.difference(current_value, prior_value), dm.abs_decimal(prior_value), decimal_places)


def calculate_cagr(
    start_value: str,
    end_value: str,
    period_count: int,
    decimal_places: int = 2,
) -> str:
    if not dm.is_filled(start_value) or not dm.is_filled(end_value):
        return ""
    if period_count <= 0:
        return ""
    start_cmp = dm.compare(start_value, "0")
    if start_cmp is None or start_cmp <= 0:
        return ""

    ratio = dm.div(end_value, start_value, 12)
    if ratio == "":
        return ""

    ratio_num = dm.to_number(ratio)
    if ratio_num is None or ratio_num <= 0:
        return ""

    cagr = (math.pow(ratio_num, 1 / period_count) - 1) * 100
    if not math.isfinite(cagr):
        return ""
    parsed = dm.parse_decimal(str(cagr))
    if parsed is None:
        return ""
    from decimal import ROUND_HALF_UP, Decimal

    quant = Decimal(1).scaleb(-decimal_places)
    return format(parsed.quantize(quant, rounding=ROUND_HALF_UP), f".{decimal_places}f")


def reconcile_reported_vs_calculated_cagr(
    reported: str,
    calculated: str,
    tolerance: str = "0.5",
) -> dict[str, Any]:
    has_reported = dm.is_filled(reported)
    has_calculated = dm.is_filled(calculated)

    if not has_reported and not has_calculated:
        return {
            "reported": "",
            "calculated": "",
            "difference": "",
            "reconciles": False,
            "message": "Neither reported nor calculated CAGR is available.",
        }

    if not has_reported:
        return {
            "reported": "",
            "calculated": calculated,
            "difference": "",
            "reconciles": False,
            "message": "Reported CAGR not entered — calculated value shown for reference only.",
        }

    if not has_calculated:
        return {
            "reported": reported,
            "calculated": "",
            "difference": "",
            "reconciles": False,
            "message": "Calculated CAGR cannot be computed from available period values.",
        }

    delta = dm.round_decimal(dm.difference(reported, calculated), 2)
    reconciles = not dm.differs_beyond(reported, calculated, tolerance)

    return {
        "reported": reported,
        "calculated": calculated,
        "difference": delta,
        "reconciles": reconciles,
        "message": (
            "Reported and calculated CAGR reconcile within tolerance."
            if reconciles
            else f"Reported CAGR differs from calculated by {delta} percentage points."
        ),
    }


def reconcile_segment_percentages(
    segments: list[dict[str, Any]],
    tolerance: str = "2",
) -> list[dict[str, Any]]:
    groups: dict[str, list[dict[str, Any]]] = {}

    for segment in segments:
        if not isinstance(segment, dict):
            continue
        parent = str(segment.get("parentMarketSeriesId") or "").strip()
        if not parent:
            continue
        period = str(segment.get("period") or "").strip()
        key = f"{parent}::{period}"
        groups.setdefault(key, []).append(segment)

    results: list[dict[str, Any]] = []

    for key, group in groups.items():
        parent_market_series_id, period = key.split("::", 1)
        percentages = [
            str(s.get("marketSharePercentage") or "")
            for s in group
            if dm.is_filled(s.get("marketSharePercentage"))
        ]
        total_percentage = dm.sum_decimals(percentages) if percentages else ""

        flags: list[str] = []
        if any(not str(s.get("parentMarketSeriesId") or "").strip() for s in group):
            flags.append("Segment missing parent market series.")
        if any(not dm.is_filled(s.get("marketSharePercentage")) for s in group):
            flags.append("One or more segments lack a market-share percentage.")
        if dm.is_filled(total_percentage):
            total_vs_100 = dm.compare(total_percentage, "100")
            if total_vs_100 is not None:
                if total_vs_100 > 0 and not dm.differs_beyond(total_percentage, "100", tolerance):
                    flags.append(
                        f"Segment percentages sum to {total_percentage}% — exceeds 100%.",
                    )
                elif total_vs_100 > 0:
                    flags.append(
                        f"Segment percentages sum to {total_percentage}% — materially above 100%.",
                    )
                lower_bound = dm.subtract("100", tolerance)
                total_vs_lower = dm.compare(total_percentage, lower_bound)
                if total_vs_lower is not None and total_vs_lower < 0:
                    flags.append(
                        f"Segment percentages sum to {total_percentage}% — materially below 100%.",
                    )

        reconciles = (
            len(flags) == 0
            and dm.is_filled(total_percentage)
            and not dm.differs_beyond(total_percentage, "100", tolerance)
        )

        results.append(
            {
                "parentMarketSeriesId": parent_market_series_id,
                "period": period,
                "totalPercentage": total_percentage,
                "reconciles": reconciles,
                "flags": flags,
            },
        )

    return results
